sync/server.md

```diff
@@ -0,0 +1,403 @@
+
+## Architecture
+
+The Sync server is intended to be horizontally scalable.
+As such, it will us a [shared-nothing-architecture](https://en.wikipedia.org/wiki/Shared-nothing_architecture).
+In addition, the server will be developed following the [12 factor app](https://12factor.net/) pattern.
+The architecture will be kept simple, where appropriate.
+There will be a single server process which serves as the following:
+
+* Static file server for the Admin UI
+* API for managing the Data Schema and other GraphQL & Data Sync features
+* GraphQL API for Apps to send Queries & Mutations to
+* Subscriptions endpoint for Apps to connect to
+
+TODO: Updated architecture diagram based on https://docs.google.com/drawings/d/1OmMY1bbI8HDhlZ_cIirv7FYTEqRRV-Hh-g0ANtYEgBY/edit
+
+The main developer flow when using the Sync Server will be something like this:
+
+* Open the Data Sync Admin UI
+* Define the Data Schema
+* Configure Data Source(s)
+* Configure Resolvers for any Queries or Mutations
+* Update/modify the Data Schema as needed, continuously saving and trying out the Schema
+
+The last part of the flow is very important.
+This implies a fast feedback loop when developing your Data Schema.
+To ensure the feedback loop is as fast as possible, the architecture of sync will be slightly different that the proof of concept architecture.
+Both the backend for the Admin UI, and the GraphQL API that Apps talk to will be part of the 1 Data Sync server.
+This server will continously run, without any restart needed when the Data Schema changes.
+To facilitate this, the mounting of the graphql express middleware will need to be done in such a way that it always rebuilds the schema based on the latests available. In reality, the parsed schema should be cached, and only reloaded/reparsed if it has changed. Here is a sample graphql endpoint mounting in express:
+
+```
+const tracing = true
+app.use('/graphql', bodyParser.json(), function (req, res, next) {
+  const schema = require('./lib/schemaParser').parseFromFile(SCHEMA_FILE, DATA_SOURCES_FILE, RESOLVER_MAPPINGS_FILE)
+  const graphql = graphqlExpress({ schema, tracing })
+  graphql(req, res, next)
+})
+```
+
+As the server needs to store resources, such as the Data Schema, Resolver Mappings and Webhooks, a data store is required.
+The Data Sync server is intended to run in a Kubernetes environment such as Openshift.
+This allows the Data Sync server to use a ServiceAccount token to save these resources in 1 or more Secrets.
+This will avoid the need for a potentially heavyweight database, at least until something different is required.
+
+A pubsub service will be required for managing the event driven nature of Subscriptions when the Data Sync server is scaled beyond 1 Replica.
+A reasonable candidate for this is Redis and its set of [pubsub commands](https://redis.io/topics/pubsub).
```

Comment:
This is still TBD.
One approach I'm thinking of is using postgres and https://github.com/voxpelli/node-pg-pubsub.
Postgres could then be the store for the Data Schema, Resolver Mappings, Data Sources, Webhooks etc...

Reply:
I've also done a little bit of searching here. It looks like MongoDB also has some pubsub capabilities using a feature called tailable cursors. The following blogpost has a high level overview with examples written in Node.

http://tugdualgrall.blogspot.com/2015/01/how-to-create-pubsub-application-with.html

Just throwing it out there as an option.

Reply:
@johnfriz this is what i'd like to discuss with you in person tomorrow.
For subscriptions in a horizontally scaled environment where Clients are connected to different instances of the server, we need a pub/sub mechanism for event notification between the server instances.

Approaches discussed so far:

* zeromq for in-memory queues, but this gets hairy when you scale up to many server instances as each instance has to connect to every other instance
* redis pub/sub https://redis.io/topics/pubsub
* AMQ
* Postgres NOTIFY/LISTEN https://github.com/voxpelli/node-pg-pubsub, https://www.postgresql.org/docs/9.2/static/sql-notify.html, https://www.postgresql.org/docs/9.2/static/sql-listen.html
* etcd (watch)
* rethinkdb

Separately, there were also some discussions about where to store the data schema, resolver mappings, webhooks etc...
In the early architecture these were going to be stored in a Secret in openshift, then mounted into the graphql server.
However, I'm questioning this approach now that the architecture is simplified (see diagram in PR) with hot-reload of the schema as its changed (no redeployment necessary)
It puts a hard dependency of k8s/openshift on the server part, which wasn't the case before.
A few things that were discussed for this:

* just store them in a secret if the server is running in OpenShift. If it's not, load/save to a file on disk (e.g. for easy local dev)
* store them in a db
  * etcd (may also be used to provide pubsub)
  * postgres (may also be used to provide pubsub)
  * rethinkdb (may also be used to provide pubsub)